File: D-MASH/client/backend/recipient_payload_v4.py
"""RecipientEnvelopeV2 and indistinguishable-at-the-schema-level cover boxes."""
import base64
import re
import secrets
from nacl.exceptions import CryptoError
from nacl.public import PrivateKey
from .route_discovery_v4 import seal,open_box


def seal_payload(recipient_public,payload):
    if not isinstance(payload,str):raise ValueError('Opaque Account payload must be a string')
    return seal(bytes(recipient_public),dict(type='RECIPIENT_PAYLOAD',version=2,
        packet_id=secrets.token_hex(32),payload=payload))


def open_payload(private_keys,blob):
    # Missing keys may mean lock/migration, never evidence of invalid ciphertext.
    if private_keys is None or (isinstance(private_keys,(tuple,list)) and not private_keys):return {'status':'deferred'}
    if not isinstance(private_keys,(tuple,list)) or len(private_keys)>2:raise ValueError('Invalid recipient key set')
    keys=[]
    for key in private_keys:
        if not isinstance(key,(bytes,bytearray,PrivateKey)):raise ValueError('Invalid recipient key')
        raw=bytes(key)
        if len(raw)!=32:raise ValueError('Invalid recipient key')
        keys.append(raw)
    for key in keys:
        try:value=open_box(key,blob)
        except (ValueError,CryptoError):continue
        if (isinstance(value,dict) and set(value)=={'type','version','packet_id','payload'}
                and value['type']=='RECIPIENT_PAYLOAD' and type(value['version']) is int and value['version']==2
                and isinstance(value['packet_id'],str) and re.fullmatch('[0-9a-f]{64}',value['packet_id'])
                and isinstance(value['payload'],str)):
            return {'status':'accepted','packet_id':value['packet_id'],'payload':value['payload']}
    return {'status':'discard'}


def cover_box(size=1024):
    if type(size) is not int or not 256<=size<=16384:raise ValueError('Invalid cover size')
    # Random raw 32-byte strings could have a distinguishable X25519 encoding.
    ephemeral=PrivateKey.generate()
    raw=bytes(ephemeral.public_key)+secrets.token_bytes(24)+secrets.token_bytes(size-56)
    return base64.b64encode(raw).decode()

File: D-MASH PWA/not_messenger/js/recipient_payload_v4.js
'use strict';
(function(global){
 const hex=value=>Array.from(value,b=>b.toString(16).padStart(2,'0')).join('');
 function sealPayload(recipientPublic,payload){
  if(typeof payload!=='string')throw Error('Opaque Account payload must be a string');
  return global.DmashRouteDiscoveryV4.seal(recipientPublic,{type:'RECIPIENT_PAYLOAD',version:2,packet_id:hex(crypto.getRandomValues(new Uint8Array(32))),payload});
 }
 function openPayload(privateKeys,blob){
  if(privateKeys==null||(Array.isArray(privateKeys)&&privateKeys.length===0))return {status:'deferred'};
  if(!Array.isArray(privateKeys)||privateKeys.length>2||privateKeys.some(key=>!(key instanceof Uint8Array)||key.length!==32))throw Error('Invalid recipient key set');
  if(!global.nacl?.box?.open||!global.DmashRouteDiscoveryV4)throw Error('Recipient crypto unavailable');
  for(const key of privateKeys){
   let value;try{value=global.DmashRouteDiscoveryV4.openBox(key,blob);}catch(_){continue;}
   if(value&&typeof value==='object'&&!Array.isArray(value)&&Object.keys(value).sort().join(',')==='packet_id,payload,type,version'&&value.type==='RECIPIENT_PAYLOAD'&&value.version===2&&typeof value.packet_id==='string'&&/^[0-9a-f]{64}$/.test(value.packet_id)&&typeof value.payload==='string')return {status:'accepted',packet_id:value.packet_id,payload:value.payload};
  }
  return {status:'discard'};
 }
 function coverBox(size=1024){
  if(!Number.isInteger(size)||size<256||size>16384)throw Error('Invalid cover size');
  const ephemeral=global.nacl.box.keyPair();
  try{const bytes=crypto.getRandomValues(new Uint8Array(size));bytes.set(ephemeral.publicKey);return btoa(String.fromCharCode(...bytes));}
  finally{ephemeral.secretKey.fill(0);}
 }
 const api=Object.freeze({sealPayload,openPayload,coverBox});global.DmashRecipientPayloadV4=api;
 if(typeof module!=='undefined')module.exports=api;
})(typeof window!=='undefined'?window:globalThis);
